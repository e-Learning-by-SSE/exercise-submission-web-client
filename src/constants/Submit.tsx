export const SubmitState = {
    DROPFILE: "Dropfile",
    FILETREE: "Filetree",
    SELECTASSIGNMENT: "SelectAssignment",
    UPLOAD: "Upload",
    RESULT: "Result"
}