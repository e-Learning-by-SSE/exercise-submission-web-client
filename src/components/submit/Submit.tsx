import React from "react";
import {Button , Segment, Header, Icon, Menu, Divider} from "semantic-ui-react";
import { SubmitState } from "../../constants/Submit";
import TreeView from "./TreeView";
import { Zip } from "../../util/ZipHelper";
import JSZip from "jszip";
import Stack from "../../stacks/Stack";
import { SubmitStack } from "../../stacks/SubmitStack";
import ErrorPortal from "../../portals/ErrorPortal";
import SubmitModal from "./SubmitModal";
import SubmissionResultModal from "./submissionResult/SubmissionResultModal";
import { SubmissionResultDto } from "exerciseserverclientlib";



export default class Submit extends React.Component<React.PropsWithChildren<{}>, {submitState: string, error: {header: string, description:string} | null,
 filelist: JSZip.JSZipObject[] | FileList | null,sModal: any}> {

    private dropfilecode: React.ReactElement<any, string | React.JSXElementConstructor<any>> ;
    private filetreecode: React.ReactElement<any, string | React.JSXElementConstructor<any>> | null = null;

    private fileInput: HTMLInputElement | null = null;

    constructor(props: React.PropsWithChildren<{}>) {
        super(props);
        this.state = {submitState: SubmitState.DROPFILE,
            error: null,
            filelist: null,
            sModal: null};
        this.dropfilecode = this.getDropFileCode();
        
    }

    private getDropFileCode(): React.ReactElement<any, string | React.JSXElementConstructor<any>> {
        return (
        <div onDrop={this.fileChange} onDragOver={this.allowDrop}>
            <Segment basic textAlign='center'>
                <Segment placeholder>
                    <Header icon>
                    <Icon name='file code outline' />
                    Drop a Zip File or Files
                    </Header>
                </Segment>
                <Divider horizontal>Or</Divider>
                <Button
                    htmlFor="folderinput"
                    primary
                    icon="folder open"
                    content="Select Folder"
                    labelPosition="right"
                    onClick={() => {this.onTriggerInputFolder()}}
                />
                {/* @ts-expect-error */}
                <input hidden id="folderinput" onChange={this.onSelectFolder} type="file" webkitdirectory="true" className="folderinput_class" ref={fileInput => this.fileInput = fileInput}></input>
            </Segment>
        </div>
        );

    }

    private getFileTreeCode(filelist: JSZip.JSZipObject[] | FileList): React.ReactElement<any, string | React.JSXElementConstructor<any>> {
        return (
            <div className="file-tree-component">
                <div className="upload-button">
                    {this.getUploadButton()}
                </div>
                <TreeView filelist={filelist}/>
            </div>
        );
    }

    private getUploadButton(): React.ReactElement<any, string | React.JSXElementConstructor<any>> {
        return (
            <div className="upload-button">
                <Menu secondary>
                    <Menu.Item>
                        <Button animated icon color="red" onClick={() => {this.setState({submitState: SubmitState.DROPFILE})}}>
                            <Button.Content visible><Icon name="refresh"></Icon>Retry</Button.Content>
                            <Button.Content hidden><Icon name="refresh"></Icon></Button.Content>
                        </Button>
                    </Menu.Item>
                    <Menu.Item position="right" >
                        <Button animated icon onClick={this.onUpload}>
                            <Button.Content visible><Icon name="upload"></Icon>Upload</Button.Content>
                            <Button.Content hidden><Icon name="upload"></Icon></Button.Content>
                        </Button>
                    </Menu.Item>
                </Menu>
            </div>);
        
    }

    onFinishedUploading=(result: SubmissionResultDto) => {
        this.setState({submitState: SubmitState.RESULT});
        this.setState({sModal: <SubmissionResultModal result={result} onClosed={() => {this.setState({sModal: null, submitState: SubmitState.DROPFILE})}}/>});
    }

    onUpload = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        event.preventDefault();
        this.setState({submitState: SubmitState.UPLOAD});
    }

    onSelectFolder = (event: React.ChangeEvent<HTMLInputElement>) => {
        event.preventDefault();
        this.setState({submitState: SubmitState.FILETREE});
        let filelist = event.target.files;
        if(filelist !== null) {
            this.filetreecode = this.getFileTreeCode(filelist);
            this.setState({submitState: SubmitState.FILETREE, filelist: filelist});
        }
    }

    onTriggerInputFolder =() => {
        if(this.fileInput !== null) {

            this.fileInput.click();
        }
    }
    
  

    fileChange = (event: React.DragEvent<HTMLDivElement>) => {
        event.preventDefault();
        let filelist = event.dataTransfer.files;
        if(filelist.length === 1 && filelist[0].name.endsWith(".zip")) {
           
            let zip = new Zip(filelist[0]).getFiles();
    
            zip.then((filelist) => {
                this.filetreecode = this.getFileTreeCode(filelist);
                this.setState({submitState: SubmitState.FILETREE, filelist: filelist});
            });

        } else if(filelist.length > 1 && filelist[0].name.endsWith(".zip")) {
            this.setState({error: {header: "Invalid Files", description: "Dont drop more than one Zip File"}});
        } else {
            this.filetreecode = this.getFileTreeCode(filelist);
            this.setState({submitState: SubmitState.FILETREE, filelist: filelist});
        }
           
    }

    allowDrop = (event: React.DragEvent<HTMLDivElement>) => {
        event.preventDefault();
    }

    
    /*upload data button with drag and drop field*/
    render() {
        return (
            <div className="submit">
                {this.state.submitState=== SubmitState.DROPFILE ? this.dropfilecode :
                 this.state.submitState=== SubmitState.FILETREE ? this.filetreecode : 
                 this.state.submitState=== SubmitState.UPLOAD ? <SubmitModal onClosed= {this.onFinishedUploading} files={this.state.filelist}/> : null}
                {this.state.sModal}
                {this.state.error ? <ErrorPortal error={this.state.error} onReady={() =>{this.setState({submitState: SubmitState.DROPFILE})}}/>: null}
            <Stack stack={SubmitStack} selected={this.state.submitState} />
            </div>
        );
    }


}

