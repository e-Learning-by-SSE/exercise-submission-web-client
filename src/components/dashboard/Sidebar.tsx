/* Sidebar in Semantic ui */

import React, { Component } from "react";
import { Menu, Icon, Sidebar, Segment } from "semantic-ui-react";
import Submit from "../submit/Submit";
import { Link } from "react-router-dom";
import DashboardContent from "./DashboardContent";

export default class SidebarComponent extends Component {

    render() {
        return (
            <Sidebar.Pushable as={Segment}>
            <Sidebar
              as={Menu}
              animation='uncover'
              icon='labeled'
              inverted
              vertical
              visible
              width='thin'
            >
              <Menu.Item as='a'>
                <Icon name='upload' />
                Submit
              </Menu.Item>
              <Menu.Item as='a'>
                <Icon name='gamepad' />
                Games
              </Menu.Item>
              <Menu.Item as='a'>
                <Icon name='camera' />
                Channels
              </Menu.Item>
            </Sidebar>
  
            <Sidebar.Pusher>
              <Segment basic>
                
                <DashboardContent>
                    <Submit/>
                </DashboardContent>
                
               
              </Segment>
            </Sidebar.Pusher>
          </Sidebar.Pushable>
        );
    }
    }