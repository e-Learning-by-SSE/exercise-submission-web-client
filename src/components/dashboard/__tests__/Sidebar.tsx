import renderer from "react-test-renderer";
import Sidebar from "../Sidebar";
import { BrowserRouter as Router } from "react-router-dom";
import Submit from "../../submit/Submit";

describe ("render Sidebar", () => {
    
        it("with sidebar visible", () => {
    
            const sidebar = renderer.create( <Router><Sidebar menuSelected={<Submit/>} sidebarVisible={true} /></Router>).toJSON();
    
            expect(sidebar).toMatchSnapshot();
        });
        it("without sidebar visible", () => {
    
            const sidebar = renderer.create(<Router><Sidebar menuSelected={<Submit/>} sidebarVisible={false} /></Router>).toJSON();
    
            expect(sidebar).toMatchSnapshot();
        });

    });