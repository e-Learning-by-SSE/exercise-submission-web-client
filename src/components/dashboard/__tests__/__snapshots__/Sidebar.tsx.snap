// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Sidebar with sidebar visible 1`] = `
<div
  className="ui segment pushable"
>
  <div
    className="ui inverted vertical labeled icon ui uncover left thin visible sidebar menu"
  >
    <a
      className="item"
      href="/submit"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="upload icon"
        onClick={[Function]}
      />
      Submit
    </a>
    <a
      className="item"
      href="/version"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="file alternate icon"
        onClick={[Function]}
      />
      Version-History
    </a>
    <a
      className="item"
      href="/settings"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="settings icon"
        onClick={[Function]}
      />
      Settings
    </a>
    <a
      className="item"
      href="https://www.uni-hildesheim.de/datenschutz/"
      onClick={[Function]}
    >
      Datenschutz
    </a>
    <a
      className="item"
      href="https://www.uni-hildesheim.de/impressum/"
      onClick={[Function]}
    >
      Impressum
    </a>
  </div>
  <div
    className="pusher"
  >
    <div
      className="ui basic placeholder segment"
    >
      <div
        className="dashboard-state-handler"
      >
        <div
          className="dashboard-content"
        >
          <div
            className="submit"
          >
            <div
              className="sequence-step"
            >
              <div
                atached="top"
                className="ui mini fluid steps"
                width={5}
              >
                <div
                  className="active step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      Drop file
                    </div>
                  </div>
                </div>
                <div
                  className="step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      View Files
                    </div>
                  </div>
                </div>
                <div
                  className="step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      Select Assignment
                    </div>
                  </div>
                </div>
                <div
                  className="step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      Upload
                    </div>
                  </div>
                </div>
                <div
                  className="step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      Result
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="ui divider"
              />
            </div>
            <div
              onDragOver={[Function]}
              onDrop={[Function]}
            >
              <div
                className="ui basic center aligned segment"
              >
                <div
                  className="ui placeholder segment"
                >
                  <div
                    className="ui icon header"
                  >
                    <i
                      aria-hidden="true"
                      className="file code outline icon"
                      onClick={[Function]}
                    />
                    Drop a Zip File or Files
                  </div>
                </div>
                <div
                  className="ui horizontal divider"
                >
                  Or
                </div>
                <button
                  className="ui icon primary right labeled button"
                  htmlFor="folderinput"
                  onClick={[Function]}
                >
                  <i
                    aria-hidden="true"
                    className="folder open icon"
                    onClick={[Function]}
                  />
                  Select Folder
                </button>
                <input
                  className="folderinput_class"
                  hidden={true}
                  id="folderinput"
                  onChange={[Function]}
                  type="file"
                  webkitdirectory="true"
                />
              </div>
            </div>
            <div
              className="ui divider"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render Sidebar without sidebar visible 1`] = `
<div
  className="ui segment pushable"
>
  <div
    className="ui inverted vertical labeled icon ui uncover left thin sidebar menu"
  >
    <a
      className="item"
      href="/submit"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="upload icon"
        onClick={[Function]}
      />
      Submit
    </a>
    <a
      className="item"
      href="/version"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="file alternate icon"
        onClick={[Function]}
      />
      Version-History
    </a>
    <a
      className="item"
      href="/settings"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="settings icon"
        onClick={[Function]}
      />
      Settings
    </a>
    <a
      className="item"
      href="https://www.uni-hildesheim.de/datenschutz/"
      onClick={[Function]}
    >
      Datenschutz
    </a>
    <a
      className="item"
      href="https://www.uni-hildesheim.de/impressum/"
      onClick={[Function]}
    >
      Impressum
    </a>
  </div>
  <div
    className="pusher"
  >
    <div
      className="ui basic placeholder segment"
    >
      <div
        className="dashboard-state-handler"
      >
        <div
          className="dashboard-content"
        >
          <div
            className="submit"
          >
            <div
              className="sequence-step"
            >
              <div
                atached="top"
                className="ui mini fluid steps"
                width={5}
              >
                <div
                  className="active step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      Drop file
                    </div>
                  </div>
                </div>
                <div
                  className="step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      View Files
                    </div>
                  </div>
                </div>
                <div
                  className="step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      Select Assignment
                    </div>
                  </div>
                </div>
                <div
                  className="step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      Upload
                    </div>
                  </div>
                </div>
                <div
                  className="step"
                  onClick={[Function]}
                >
                  <div
                    className="content"
                  >
                    <div
                      className="title"
                    >
                      Result
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="ui divider"
              />
            </div>
            <div
              onDragOver={[Function]}
              onDrop={[Function]}
            >
              <div
                className="ui basic center aligned segment"
              >
                <div
                  className="ui placeholder segment"
                >
                  <div
                    className="ui icon header"
                  >
                    <i
                      aria-hidden="true"
                      className="file code outline icon"
                      onClick={[Function]}
                    />
                    Drop a Zip File or Files
                  </div>
                </div>
                <div
                  className="ui horizontal divider"
                >
                  Or
                </div>
                <button
                  className="ui icon primary right labeled button"
                  htmlFor="folderinput"
                  onClick={[Function]}
                >
                  <i
                    aria-hidden="true"
                    className="folder open icon"
                    onClick={[Function]}
                  />
                  Select Folder
                </button>
                <input
                  className="folderinput_class"
                  hidden={true}
                  id="folderinput"
                  onChange={[Function]}
                  type="file"
                  webkitdirectory="true"
                />
              </div>
            </div>
            <div
              className="ui divider"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
